src/lib/crypto/crypto.c

```diff
@@ -406,22 +421,46 @@ crypto_EVP_get_digestbyname(const char *name)
 }
 
 int
-crypto_HMAC_Init_ex(HMAC_CTX *ctx, const void *key, int len, const EVP_MD *md,
-		    ENGINE *impl)
+crypto_HMAC_Init_ex(crypto_HMAC_CTX *ctx, const void *key, int len,
+		    const char *digest, const EVP_MD *md, ENGINE *impl)
 {
+#if OPENSSL_VERSION_NUMBER < 0x30000000L
+	(void)digest;
 	return HMAC_Init_ex(ctx, key, len, md, impl);
+#else
+	(void)md;
+	(void)impl;
+	OSSL_PARAM params[2];
+	params[0] = OSSL_PARAM_construct_utf8_string("digest",
+						     (char *)digest, 0);
+	params[1] = OSSL_PARAM_construct_end();
```

Comment:
1. Shouldn't we use `OSSL_PARAM_set_octet_string()`? We can pass any binary data from Lua here.
2. What if I'll pass data with zero in the middle?

Reply:
I misinterpreted a purpose of `digest` argument. Now I got it, it is "md5", "sha1" and so on. Sorry for noise.